server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -279,5 +225,59 @@ public class Response {
                 }
             }
         }
+
+        void finish() {
+            Map<String, Map<Integer, List<StoreStatus>>> indicesStatuses = new HashMap<>();
+            List<Failure> failures = new ArrayList<>();
+            for (InternalAsyncFetch.Response fetchResponse : fetchResponses) {
+                var indexName = fetchResponse.shardId.getIndexName();
+                var shardId = fetchResponse.shardId.id();
+                var indexStatuses = indicesStatuses.computeIfAbsent(indexName, k -> new HashMap<>());
+                var storeStatuses = indexStatuses.computeIfAbsent(shardId, k -> new ArrayList<>());
+
+                for (NodeGatewayStartedShards r : fetchResponse.responses) {
+                    if (shardExistsInNode(r)) {
+                        var allocationStatus = getAllocationStatus(indexName, shardId, r.getNode());
+                        storeStatuses.add(new StoreStatus(r.getNode(), r.allocationId(), allocationStatus, r.storeException()));
+                    }
+                }
+
+                for (FailedNodeException failure : fetchResponse.failures) {
+                    failures.add(new Failure(failure.nodeId(), indexName, shardId, failure.getCause()));
+                }
+            }
+            // make the status structure immutable
+            indicesStatuses.replaceAll((k, v) -> {
+                v.replaceAll((s, l) -> {
+                    CollectionUtil.timSort(l);
+                    return List.copyOf(l);
+                });
+                return Map.copyOf(v);
+            });
+            listener.onResponse(new IndicesShardStoresResponse(Map.copyOf(indicesStatuses), List.copyOf(failures)));
+        }
+
+        private AllocationStatus getAllocationStatus(String index, int shardID, DiscoveryNode node) {
```

Comment:
This method had no business being a member of `InternalAsyncFetch`, this change just moves it up a level.